File: 백준/Silver/1012. 유기농 배추/유기농 배추.py
from collections import deque

test = int(input())

dx = [0,0,1,-1]
dy = [1,-1,0,0]

def bfs(graph, a, b):
    q = deque()
    # 방문한 구역 0으로 처리
    q.append((a,b))
    graph[a][b] = 0
    # 근처에 1이 없을 때까지 반복
    while q:
        x, y = q.popleft()
        # 상하좌우 탐색
        for i in range(4):
            nx = x + dx[i]
            ny = y + dy[i]
            # 구역을 벗어난 경우 건너뛰기
            if nx < 0 or nx >= n or ny < 0 or ny >= m:
                continue
            # 근처의 1들을 0으로 초기화
            if graph[nx][ny] == 1:
                graph[nx][ny] = 0
                q.append((nx, ny))


for i in range(test):
    cnt = 0
    n, m, k = map(int, input().split())
    # 그래프 생성
    graph = [[0] * m for _ in range(n)]

    # 배추 위치 표시
    for j in range(k):
        x, y = map(int, input().split())
        graph[x][y] = 1
    
    # 완전탐색하여 배추 위치 찾기
    for a in range(n):
        for b in range(m):
            # 배추를 찾음
            if graph[a][b] == 1:
                bfs(graph, a,b)
                cnt += 1
    
    print(cnt)

File: 백준/Silver/1012. 유기농 배추/유기농 배추.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.*;

public class Main {
	static int[][] map;
	static int n;
	static int m;
	static int k;
	static boolean[][] visited;

	public static void main(String[] args) throws IOException {
		BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
		int t = Integer.parseInt(br.readLine());
		for(int test = 1; test<=t;test++) {
			int cnt = 0;
			StringTokenizer st = new StringTokenizer(br.readLine());
			n = Integer.parseInt(st.nextToken());
			m = Integer.parseInt(st.nextToken());
			k = Integer.parseInt(st.nextToken());
			map = new int[n][m];
			for(int i = 0;i<k;i++) {
				st = new StringTokenizer(br.readLine());
				int x = Integer.parseInt(st.nextToken());
				int y = Integer.parseInt(st.nextToken());
				map[x][y] = 1; // 배추 위치 표시
				
				
			}
		
			visited = new boolean[n][m];
			// 배추 탐색
			for(int i = 0; i<n;i++) {
				for(int j = 0; j<m;j++) {
					//배추가 있고 방문하지 않은 배추일 경우
					if(map[i][j] == 1 && !visited[i][j]) {
						bfs(i,j);
						cnt++;
					}
				}
			}
			System.out.println(cnt);
		}

	}
	static int[] dx = {0,0,1,-1};
	static int[] dy = {1,-1,0,0};
 	static void bfs(int x, int y) {
		Integer[] xy = new Integer[2];
		xy[0] = x;
		xy[1] = y;
		Queue<Integer[]> q = new ArrayDeque();
		q.add(xy);
		visited[x][y] = true;
		while(!q.isEmpty()) {
			Integer[] cicj = q.poll();
			for(int d =0; d<4;d++) {
				int di = cicj[0] + dx[d];
				int dj = cicj[1] + dy[d];
				 if(di <0 || di >= n || dj <0 || dj>=m) continue;
				 
				 if(!visited[di][dj] && map[di][dj] == 1) {
					 visited[di][dj] = true;
					 
					 q.add(new Integer[] {di, dj});
				 }
			}
		}
	}

}